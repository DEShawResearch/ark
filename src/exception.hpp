#ifndef ark_exception_hpp
#define ark_exception_hpp

#include <stdexcept>

/*! \file ark/exception.hpp
  Ark exception handling 
*/

namespace Ark {

  //! Ark::exception inherits from std::runtime_error.
  struct exception : public std::runtime_error {
    /*! Constructor.
      @param s the exception message.
    */
    explicit exception(const std::string& s);
  };

  //! Ark::InputError is an alternative which can be thrown by parsers
  //! for invalid inputs.
  struct InputError : public exception {

    /*! Constructor.
      @param s the exception message.
    */
    explicit InputError(const std::string& s) : exception(s) {}

    /*! Constructor to allow coercion of regular exceptions into InputErrors.
      @param e an Ark::exception
    */
    explicit InputError(const exception& e) : exception(e) {}
  };

}

#endif
