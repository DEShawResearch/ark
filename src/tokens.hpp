#ifndef ark_tokens_hpp
#define ark_tokens_hpp

#include <string>
#include <iostream>

/*! \file ark/tokens.hpp

One of the auxiliary classes used by Ark is a simple character-based
tokenizer for parsing inputs and queries.
*/

namespace Ark {
  /*! A token is a variadic type representing a chunk of input, delimited
   by "syntax characters" or a single "syntax character". */
  class token {
    friend class tokenizer;
  public:
    //! The things a token might be.
    enum kind_t {
      None=0, //!< Used to initialize.
      End,    //!< End of input stream.
      Syntax, //!< A syntax character.
      Symbol, //!< A maximal non-whitespace, non-syntax substring of input.
      String  //!< A quote (special syntax characters) delimited substring.
    };
  private:
    kind_t k; //!< The kind of this token.
    std::string buf; //!< A temporary buffer for input.

  public:
    //! default initialized the kind to None.
    token() : k(None), buf("") {}
    /*!
      returns the kind.
      @return the kind of this token.
    */
    kind_t      kind() const {return k;}

    /*! Return the syntax character of this token or char(0) if this is
      a non-syntax token.
      @return syntax character.
    */
    char syntax() const {return (k==Syntax) ? (buf[0]) : ('\0');}
    /*! Return the text of this token as a C++-string.
      @return the text of this token.
    */
    std::string text() const;
  };


  /*! A tokenizer produces a stream of tokens from a C++ input stream.
    The interface is a basic iteration scheme supporting a call like
    <code>
\verbatim
    for(tokenizer t(cin,syn); t.next().current() != End; ) {
      switch(t.current().kind()) {....}
    }
\endverbatim
    </code>
  */
  class tokenizer {
  public:
    //! Syntax information is held in an internal class.
    class syntax {
      // syntax information
      std::string comment; //!< The start of comment character.  Comments
                           //!< currently start from a comment character and
                           //!< go until a newline.  Might change this some
                           //!< day.
      std::string syntax_chars; //!< syntax characters (non-quote).
      std::string quote; //!< allowed quote characters for string.
      std::string reserved; //!< syntax+quote+comment concatenation.
    public:
      /*! Constructor.
        @param syn syntax characters.
        @param com comment characters.

        Currently quote characters are all of "'`, hard-coded.
      */
      syntax(const std::string &syn,const std::string &com):
        comment(com), syntax_chars(syn), quote("\"'`"),
        reserved(comment+syntax_chars+quote)
      {}
      //! test for comment character.
      //! @param c
      //! @return whether a comment.
      bool is_comment(char c) const {
        return std::string::npos != comment.find(c);
      }
      //! test for syntax character.
      //! @param c
      //! @return whether a syntax.
      bool is_syntax(char c) const {
        return std::string::npos != syntax_chars.find(c);
      }
      //! test for quote character.
      //! @param c
      //! @return whether a quote.
      bool is_quote(char c) const {
        return std::string::npos != quote.find(c);
      }
      //! test for reserved character.
      //! @param c
      //! @return whether a reserved.
      bool is_reserved(char c) const {
        return std::string::npos != reserved.find(c);
      }
    };

  private:
    syntax S;
    // tokenizing state
    std::istream &in;
    unsigned      line,col;
    token         t;

    void nextch(void); // yes this BS does help.  C++ iostreams blow.
    static const size_t BUFsize=256;
    char *C,*E,BUF[BUFsize];

  public:
    /*! Construct a tokenizer from an input stream and a syntax.
      @param i input stream.
      @param syn syntax.

      Initializes some internal meta-information about the stream to
      track line numbers and such.

      The initial value of current() is None.
    */
    tokenizer(std::istream &i,syntax syn)
      : S(syn),in(i),line(1),col(0),t() {
      C=BUF;
      in.get(C,BUFsize,0);
      E=C+in.gcount();
    }
    //! The current line number in the stream.
    //! @return the line number.
    unsigned lineno() const {return line;}
    //! The current column number in the stream.
    //! @return the column number.
    unsigned colno()  const {return col; }

    //! The current token in the stream.
    //! @return a token reference.
    const token &current() const {return t;}

    //! The next token in the stream according to a given syntax.
    //! @param s given syntax.
    //! @return token reference.
    const token &next(const syntax &s);
    //! The next token in the stream according the syntax from the
    //! constructor.
    //! @return token reference.
    inline const token &next() { return next(S); }
  };
}

#endif
