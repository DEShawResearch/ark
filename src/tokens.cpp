#include "exception.hpp"
#include "tokens.hpp"
#include <sstream>
#include <ctype.h>

namespace Ark {
  std::string token::text() const {
    switch(k) {
    case None:    return "<undefined>"; 
    case End:     return "<end of file>"; 
    case Syntax:
    case Symbol:  return buf;
    case String:  return buf;//.substr(1,buf.size()-2);
    }
    // can't reach here
    return "no way I reached this line here";
  }

  void tokenizer::nextch(void) {
    if (*C=='\n') ++line, col=0; else ++col;
    if (++C==E) {
      C=BUF;
      in.get(C,BUFsize,0);
      E=C+in.gcount();
    }
  }

  const token &tokenizer::next(const syntax &S) {
    // mop up whitespace and comments
    // ok so current char is non-whitespace within bounds
    while( C!=E ) {
      while( C!=E && isspace(*C) ) nextch();
      if (S.is_comment(*C))
        while( C!=E && *C!='\n' ) nextch();
      else break;
    }
    if (C==E) {
      t.k   = token::End;
      nextch();
    }
    else if (S.is_syntax(*C)) {
      t.k   = token::Syntax;
      t.buf = *C;
      nextch();
    }
    else if (S.is_quote(*C)) {
      t.k   = token::String;
      //t.buf = *C;
      t.buf.clear();
      char q = *C;
      bool esc=0;
      for( nextch(); C!=E && (*C!=q || esc) ; nextch() )
        if ( *C !='\\' || esc) {
          t.buf += *C; esc=0;
        }
        else esc=1;
      if (*C != q) // check terminal quote
        throw InputError("invalid string token");
      //t.buf += *C;
      nextch();
    }
    else {
      t.k   = token::Symbol;
      t.buf = *C;
      for(nextch(); C!=E && !isspace(*C) && !S.is_reserved(*C);nextch())
        t.buf += *C;
    }
    return t;
  }
}
