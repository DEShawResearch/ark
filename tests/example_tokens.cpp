#include <ark/tokens.hpp>
#include <ark/exception.hpp>

#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <fstream>

using namespace Ark;
using namespace std;

int main(int argc,char **argv) {
  if (argv[1]) {
    try {
      tokenizer::syntax S("()[]<>{}.?!=","#");
      ifstream F(argv[1]);
      for(tokenizer T(F,S); T.next().kind() !=token::End; ) {
        std::cout << T.current().text()
                  << " at line " << T.lineno()
                  << " and col " << T.colno() 
                  << " of file " << argv[1] << std::endl;
      }
    }
    catch (const Ark::exception &e) {
      fprintf(stderr,"Caught: '%s'\n",e.what());
      exit(1);
    }
  }
  else {
    fprintf(stderr,"usage: %s testfile\n",argv[0]);
    exit(1);
  }
}
