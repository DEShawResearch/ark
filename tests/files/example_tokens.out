z at line 1 and col 1 of file testfile
= at line 1 and col 2 of file testfile
5 at line 1 and col 4 of file testfile
x at line 2 and col 1 of file testfile
= at line 2 and col 2 of file testfile
[ at line 2 and col 4 of file testfile
2 at line 2 and col 5 of file testfile
3 at line 2 and col 7 of file testfile
4 at line 2 and col 9 of file testfile
5 at line 2 and col 11 of file testfile
] at line 2 and col 12 of file testfile
y at line 3 and col 1 of file testfile
= at line 3 and col 2 of file testfile
[ at line 3 and col 4 of file testfile
{ at line 3 and col 6 of file testfile
myarray at line 3 and col 14 of file testfile
= at line 3 and col 15 of file testfile
[ at line 3 and col 17 of file testfile
3 at line 3 and col 18 of file testfile
4 at line 3 and col 20 of file testfile
5 at line 3 and col 22 of file testfile
] at line 3 and col 23 of file testfile
} at line 3 and col 26 of file testfile
yuiop at line 3 and col 32 of file testfile
qwerty at line 3 and col 39 of file testfile
] at line 3 and col 40 of file testfile
t at line 4 and col 1 of file testfile
= at line 4 and col 2 of file testfile
{ at line 4 and col 4 of file testfile
e at line 4 and col 5 of file testfile
= at line 4 and col 6 of file testfile
4 at line 4 and col 7 of file testfile
multi at line 4 and col 13 of file testfile
= at line 4 and col 14 of file testfile
{ at line 4 and col 16 of file testfile
record at line 4 and col 22 of file testfile
= at line 4 and col 23 of file testfile
here at line 4 and col 30 of file testfile
! at line 4 and col 32 of file testfile
include at line 4 and col 39 of file testfile
includeme at line 4 and col 49 of file testfile
ijkl at line 5 and col 5 of file testfile
= at line 5 and col 6 of file testfile
abcd at line 5 and col 11 of file testfile
} at line 5 and col 12 of file testfile
} at line 5 and col 13 of file testfile
y at line 7 and col 1 of file testfile
[ at line 7 and col 2 of file testfile
+ at line 7 and col 3 of file testfile
] at line 7 and col 4 of file testfile
= at line 7 and col 6 of file testfile
added at line 7 and col 12 of file testfile
y at line 8 and col 1 of file testfile
[ at line 8 and col 2 of file testfile
+ at line 8 and col 3 of file testfile
] at line 8 and col 4 of file testfile
= at line 8 and col 6 of file testfile
added again at line 8 and col 20 of file testfile
x at line 10 and col 1 of file testfile
= at line 10 and col 2 of file testfile
[ at line 10 and col 4 of file testfile
{ at line 10 and col 5 of file testfile
tau at line 10 and col 8 of file testfile
= at line 10 and col 9 of file testfile
[ at line 10 and col 11 of file testfile
1 at line 10 and col 12 of file testfile
2 at line 10 and col 14 of file testfile
] at line 10 and col 15 of file testfile
} at line 10 and col 16 of file testfile
{ at line 10 and col 18 of file testfile
tau at line 10 and col 21 of file testfile
= at line 10 and col 22 of file testfile
[ at line 10 and col 24 of file testfile
23 at line 10 and col 26 of file testfile
{ at line 10 and col 28 of file testfile
hey at line 10 and col 31 of file testfile
= at line 10 and col 32 of file testfile
there at line 10 and col 40 of file testfile
} at line 10 and col 41 of file testfile
56 at line 10 and col 44 of file testfile
] at line 10 and col 45 of file testfile
} at line 10 and col 46 of file testfile
] at line 10 and col 47 of file testfile
jgak at line 12 and col 4 of file testfile
= at line 14 and col 1 of file testfile
123 at line 14 and col 4 of file testfile
file at line 16 and col 4 of file testfile
= at line 16 and col 6 of file testfile
blah blah `file` here's \`'s at line 16 and col 40 of file testfile
ggg at line 18 and col 3 of file testfile
. at line 18 and col 5 of file testfile
! at line 18 and col 7 of file testfile
include at line 18 and col 14 of file testfile
includeme at line 18 and col 24 of file testfile
nothing at line 20 and col 7 of file testfile
= at line 20 and col 9 of file testfile
[ at line 20 and col 10 of file testfile
1 at line 20 and col 11 of file testfile
? at line 20 and col 13 of file testfile
4 at line 20 and col 15 of file testfile
] at line 20 and col 16 of file testfile
array at line 22 and col 5 of file testfile
= at line 22 and col 7 of file testfile
[ at line 22 and col 8 of file testfile
1 at line 22 and col 10 of file testfile
2 at line 22 and col 12 of file testfile
3 at line 22 and col 14 of file testfile
{ at line 22 and col 16 of file testfile
} at line 22 and col 17 of file testfile
] at line 22 and col 19 of file testfile
array at line 23 and col 5 of file testfile
[ at line 23 and col 6 of file testfile
3 at line 23 and col 7 of file testfile
] at line 23 and col 8 of file testfile
= at line 23 and col 9 of file testfile
0x27 at line 23 and col 15 of file testfile
array at line 24 and col 5 of file testfile
[ at line 24 and col 6 of file testfile
4 at line 24 and col 7 of file testfile
] at line 24 and col 8 of file testfile
= at line 24 and col 9 of file testfile
010 at line 24 and col 14 of file testfile
f at line 25 and col 1 of file testfile
= at line 25 and col 2 of file testfile
[ at line 25 and col 4 of file testfile
1 at line 25 and col 5 of file testfile
] at line 25 and col 6 of file testfile
ggg at line 26 and col 3 of file testfile
. at line 26 and col 4 of file testfile
rrr at line 26 and col 7 of file testfile
[ at line 26 and col 8 of file testfile
0 at line 26 and col 9 of file testfile
] at line 26 and col 10 of file testfile
= at line 26 and col 11 of file testfile
5 at line 26 and col 13 of file testfile
ggg at line 27 and col 3 of file testfile
. at line 27 and col 4 of file testfile
rrr at line 27 and col 7 of file testfile
[ at line 27 and col 8 of file testfile
1 at line 27 and col 9 of file testfile
] at line 27 and col 10 of file testfile
= at line 27 and col 11 of file testfile
xx at line 27 and col 16 of file testfile
ggg at line 28 and col 3 of file testfile
. at line 28 and col 4 of file testfile
rrr at line 28 and col 7 of file testfile
[ at line 28 and col 8 of file testfile
2 at line 28 and col 9 of file testfile
] at line 28 and col 10 of file testfile
= at line 28 and col 11 of file testfile
xxx at line 28 and col 15 of file testfile
ggg at line 29 and col 3 of file testfile
. at line 29 and col 4 of file testfile
rrr at line 29 and col 7 of file testfile
[ at line 29 and col 8 of file testfile
1 at line 29 and col 9 of file testfile
] at line 29 and col 10 of file testfile
. at line 29 and col 11 of file testfile
jj at line 29 and col 13 of file testfile
= at line 29 and col 15 of file testfile
183 at line 29 and col 20 of file testfile
foo at line 31 and col 3 of file testfile
= at line 31 and col 5 of file testfile
{ at line 31 and col 6 of file testfile
bar at line 31 and col 10 of file testfile
= at line 31 and col 12 of file testfile
1 at line 31 and col 13 of file testfile
} at line 31 and col 15 of file testfile
foo at line 32 and col 3 of file testfile
. at line 32 and col 4 of file testfile
bar at line 32 and col 7 of file testfile
= at line 32 and col 9 of file testfile
2 at line 32 and col 10 of file testfile
foo at line 33 and col 3 of file testfile
= at line 33 and col 5 of file testfile
1 at line 33 and col 6 of file testfile
foo at line 35 and col 3 of file testfile
[ at line 35 and col 4 of file testfile
0 at line 35 and col 5 of file testfile
] at line 35 and col 6 of file testfile
= at line 35 and col 8 of file testfile
3 at line 35 and col 9 of file testfile
ert at line 36 and col 3 of file testfile
= at line 36 and col 4 of file testfile
17 at line 36 and col 6 of file testfile
A at line 38 and col 1 of file testfile
= at line 38 and col 2 of file testfile
{ at line 38 and col 3 of file testfile
x at line 38 and col 5 of file testfile
= at line 38 and col 6 of file testfile
1 at line 38 and col 7 of file testfile
} at line 38 and col 9 of file testfile
A at line 39 and col 1 of file testfile
{ at line 39 and col 3 of file testfile
y at line 39 and col 5 of file testfile
= at line 39 and col 6 of file testfile
2 at line 39 and col 7 of file testfile
} at line 39 and col 9 of file testfile
B at line 40 and col 1 of file testfile
= at line 40 and col 2 of file testfile
{ at line 40 and col 3 of file testfile
x at line 40 and col 5 of file testfile
= at line 40 and col 6 of file testfile
1 at line 40 and col 7 of file testfile
} at line 40 and col 9 of file testfile
B at line 41 and col 1 of file testfile
= at line 41 and col 2 of file testfile
{ at line 41 and col 3 of file testfile
y at line 41 and col 5 of file testfile
= at line 41 and col 6 of file testfile
2 at line 41 and col 7 of file testfile
} at line 41 and col 9 of file testfile
